fish-rust/src/event.rs

```diff
@@ -0,0 +1,926 @@
+            }
+            handlers.remove(i);
+            removed += 1;
+        }
+    }
+
+    removed
+}
+
+/// Remove all events for the given function name.
+pub fn remove_function_handlers(name: &wstr) -> usize {
+    remove_handlers_if(|h| h.function_name == name)
+}
+
+fn event_remove_function_handlers_ffi(name: &CxxWString) -> usize {
+    remove_function_handlers(name.as_wstr())
+}
+
+/// Return all event handlers for the given function.
+pub fn get_function_handlers(name: &wstr) -> EventHandlerList {
+    EVENT_HANDLERS
+        .lock()
+        .expect("event handler list should not be poisoned")
+        .iter()
+        .filter(|h| h.function_name == name)
+        .cloned()
+        .collect()
+}
+
+fn event_get_function_handler_descs_ffi(name: &CxxWString) -> Vec<event_description_t> {
+    get_function_handlers(name.as_wstr())
+        .iter()
+        .map(|h| event_description_t::from(&h.desc))
+        .collect()
+}
+
+/// Perform the specified event. Since almost all event firings will not be matched by even a single
+/// event handler, we make sure to optimize the 'no matches' path. This means that nothing is
+/// allocated/initialized unless needed.
+fn fire_internal(parser: &mut parser_t, event: &Event) {
+    assert!(
+        parser.libdata_pod().is_event >= 0,
+        "is_event should not be negative"
+    );
+
+    let saved_is_event = parser.libdata_pod().is_event;
+    parser.libdata_pod().is_event += 1;
+    // Suppress fish_trace during events.
+    let saved_suppress_fish_trace = parser.libdata_pod().suppress_fish_trace;
+    parser.libdata_pod().suppress_fish_trace = true;
+
+    // Capture the event handlers that match this event.
+    let fire: Vec<_> = EVENT_HANDLERS
+        .lock()
+        .expect("event handler list should not be poisoned")
+        .iter()
+        .filter(|h| h.matches(event))
+        .cloned()
+        .collect();
+
+    // Iterate over our list of matching events. Fire the ones that are still present.
+    let mut fired_one_shot = false;
+    for handler in fire {
+        // A previous handler may have erased this one.
+        if handler.removed.load(Ordering::Relaxed) {
+            continue;
+        };
+
+        // Construct a buffer to evaluate, starting with the function name and then all the
+        // arguments.
+        let mut buffer = handler.function_name.clone();
+        for arg in &event.arguments {
+            buffer.push(' ');
+            buffer.push_utfstr(&escape_string(
+                arg,
+                EscapeStringStyle::Script(EscapeFlags::default()),
+            ));
+        }
+
+        // Event handlers are not part of the main flow of code, so they are marked as
+        // non-interactive.
+        let saved_is_interactive = parser.libdata_pod().is_interactive;
+        parser.libdata_pod().is_interactive = false;
+        let prev_statuses = parser.get_last_statuses().within_unique_ptr();
+
+        FLOG!(
+            event,
```

Comment:
maybe we should add an incomplete definition of `FLOGF` and start using that already, fixing the definition later